azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/GenericResource.java

```diff
@@ -54,179 +54,203 @@
-         * @return the next stage of the generic resource definition
+         * A generic resource definition allowing resource group to be specified.
          */
-        DefinitionWithOrWithoutParentResource withProviderNamespace(String resourceProviderNamespace);
-    }
+        interface WithGroup extends GroupableResource.DefinitionStages.WithGroup<WithResourceType> {
+        }
 
-    /**
-     * A generic resource definition allowing parent resource to be specified.
-     */
-    interface DefinitionWithOrWithoutParentResource extends DefinitionWithPlan {
         /**
-         * Specifies the parent resource.
-         *
-         * @param parentResourceId the parent resource id
-         * @return the next stage of the generic resource definition
+         * A generic resource definition allowing resource type to be specified.
          */
-        DefinitionWithPlan withParentResource(String parentResourceId); // ParentResource is optional so user can navigate to DefinitionWithPlan with or without it.
-    }
+        interface WithResourceType {
+            /**
+             * Specifies the resource's type.
+             *
+             * @param resourceType the type of the resources
+             * @return the next stage of generic resource definition
+             */
+            WithProviderNamespace withResourceType(String resourceType);
+        }
 
-    /**
-     * A generic resource definition allowing plan to be specified.
-     */
-    interface DefinitionWithPlan {
         /**
-         * Specifies the plan of the resource. The plan can only be set for 3rd party resources.
-         *
-         * @param name the name of the plan
-         * @param publisher the publisher of the plan
-         * @param product the name of the product
-         * @param promotionCode the promotion code, if any
-         * @return the next stage of the generic resource definition
+         * A generic resource definition allowing provider namespace to be specified.
          */
-        DefinitionWithApiVersion withPlan(String name, String publisher, String product, String promotionCode);
+        interface WithProviderNamespace {
+            /**
+             * Specifies the resource provider's namespace.
+             *
+             * @param resourceProviderNamespace the namespace of the resource provider
+             * @return the next stage of the generic resource definition
+             */
+            WithOrWithoutParentResource withProviderNamespace(String resourceProviderNamespace);
+        }
 
         /**
-         * Specifies the plan of the resource.
-         *
-         * @return the next stage of the generic resource definition
+         * A generic resource definition allowing parent resource to be specified.
          */
-        DefinitionWithApiVersion withoutPlan();
-    }
+        interface WithOrWithoutParentResource extends WithPlan {
```

Comment:
Based on the comment, I think I understand the intent, but then why idn't WithPlan the next interface in the required chain  (instead of WithOrWithoutParentResource) and then this interface (renamed as WithParentResource) simply becomes one of the optionals? (WithCreate)